### PR TITLE
notify about finalized block in exex

### LINKED ISSUE #8794: ability to check if block is finalized  in Exex
### Describe the feature

It could be beneficial to get finalized updates only in via ExEx notification stream
and @mattsse 
could you provide more context?

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lakshya-sky**
Closes: #8794

### CODE REVIEW

File: crates/exex/exex/src/notification.rs [hunk 1]
```diff
@@ -22,6 +23,8 @@ pub enum ExExNotification {
         /// The old chain before reversion.
         old: Arc<Chain>,
     },
+    /// Finalized block header.
+    FinalizedBlock(Option<SealedHeader>),
```
Comment (shekhirin) [inline crates/exex/exex/src/notification.rs:27]: just to be consistent with other names

```suggestion
    BlockFinalized(Option<SealedHeader>),
```

File: crates/node/builder/src/launch/exex.rs [hunk 2]
```diff
@@ -91,15 +92,28 @@ impl<Node: FullNodeComponents + Clone> ExExLauncher<Node> {
...
+                            handle
+                                .send_async(notification.into())
+                                .await
+                                .expect("blockchain tree notification could not be sent to exex manager");
+                        }
+                        block_header = finalized_block_notifications.next() => {
+                            handle
+                                .send_async(block_header.into())
+                                .await
+                                .expect("blockchain tree notification could not be sent to exex manager");
```
Comment (shekhirin) [inline crates/node/builder/src/launch/exex.rs:113]:
```suggestion
                                .expect("blockchain tree canon state notification could not be sent to exex manager");
                        }
                        block_header = finalized_block_notifications.next() => {
                            handle
                                .send_async(block_header.into())
                                .await
                                .expect("blockchain tree finalized block notification could not be sent to exex manager");
```

File: crates/storage/provider/src/traits/finalized_block.rs [hunk 3]
```diff
@@ -12,3 +19,50 @@ pub trait FinalizedBlockWriter: Send + Sync {
     /// Saves the given finalized block number in the DB.
     fn save_finalized_block_number(&self, block_number: BlockNumber) -> ProviderResult<()>;
 }
+
+/// Type alias for a receiver that receives [`CanonStateNotification`]
```
Comment (shekhirin) [inline crates/storage/provider/src/traits/finalized_block.rs:23]: comments are outdated here and in other places in this file

File: crates/exex/exex/src/notification.rs (hunk 1 above)
Comment (shekhirin) [inline crates/exex/exex/src/notification.rs:27]: why is it `Option` and not just a `SealedHeader`?

**Review by shekhirin — CHANGES_REQUESTED** [review]
I think that's what we want but would like to defer to @mattsse

File: crates/exex/exex/src/notification.rs (hunk 1 above)
Comment (lakshya-sky) [reply crates/exex/exex/src/notification.rs:27]: you're right it should be without Option.

File: crates/exex/exex/src/manager.rs [hunk 4]
```diff
@@ -109,7 +109,9 @@ impl ExExHandle {
                 // [ExExNotification::ChainReverted] cases and always send the
                 // notification, because the ExEx should be aware of the reorgs and reverts lower
                 // than its finished height
-                ExExNotification::ChainReorged { .. } | ExExNotification::ChainReverted { .. } => {}
+                ExExNotification::ChainReorged { .. } |
+                ExExNotification::ChainReverted { .. } |
+                ExExNotification::BlockFinalized(_) => {}
```
Comment (shekhirin) [inline crates/exex/exex/src/manager.rs:114]: hmm, I think this variant should be in the arm above where we handle `ExExNotification::ChainCommitted`. @onbjerg wdyt?

File: crates/node/builder/src/launch/exex.rs [hunk 5]
```diff
@@ -91,15 +92,30 @@ impl<Node: FullNodeComponents + Clone> ExExLauncher<Node> {
...
+                                .expect("blockchain tree canonical state notification could not be sent to exex manager");
+                        }
+                        block_header = finalized_block_notifications.next() => {
+                            if let Some(notification) = block_header{
+                                handle
+                                    .send_async(notification.into())
+                                    .await
+                                    .expect("blockchain tree finalized block notification could not be sent to exex manager");
+                            }
+                        }
```
Comment (shekhirin) [inline crates/node/builder/src/launch/exex.rs:116]: > `Poll::Ready(None)` means that the stream has terminated, and poll_next should not be invoked again.

from https://docs.rs/futures-core/0.3.30/futures_core/stream/trait.Stream.html#tymethod.poll_next

```suggestion
                        Some(block_header) = finalized_block_notifications.next() => {
                            handle
                                .send_async(notification.into())
                                .await
                                .expect("blockchain tree finalized block notification could not be sent to exex manager");
                        }
```

**Review by shekhirin — CHANGES_REQUESTED** [review]

File: crates/exex/exex/src/manager.rs (hunk 4 above)
Comment (lakshya-sky) [reply crates/exex/exex/src/manager.rs:114]: I am thinking here out loud that would there be a case where we receive a notification about a finalized block with lower height than an already finalized block?

**Review by mattsse — CHANGES_REQUESTED** [review]
this pr does two things

adding/changing the finalized block tracker to a watch and integrate it in exex

this pr should only do the former and we do the latter separately because the latter is more complex.

so please submit the watch channel feature as a standalone pr first.

**Pull request closed without merging**